kvbc/src/v4blockchain/detail/blockchain.cpp

```diff
@@ -64,6 +64,44 @@ BlockId Blockchain::addBlock(const concord::kvbc::categorization::Updates& categ
   return id;
 }
 
+// Delete up to until not including until,
+// returns the last block id that was deleted.
+BlockId Blockchain::deleteBlocksUntil(BlockId until) {
+  ConcordAssertGT(genesis_block_id_, INVALID_BLOCK_ID);
+  ConcordAssertLT(genesis_block_id_, until);
+  // We have a single block on the chain
+  if (last_reachable_block_id_ == genesis_block_id_) return genesis_block_id_ - 1;
+  // We don't want to erase all the blockchain
+  const auto last_deleted_block = std::min(last_reachable_block_id_.load() - 1, until - 1);
+  auto write_batch = native_client_->getBatch();
+  for (uint64_t i = genesis_block_id_; i <= last_deleted_block; ++i) {
+    deleteBlock(i, write_batch);
+  }
+  native_client_->write(std::move(write_batch));
+  auto blocks_deleted = (last_deleted_block - genesis_block_id_) + 1;
+  genesis_block_id_ = last_deleted_block + 1;
+
+  LOG_INFO(V4_BLOCK_LOG, "Deleted " << blocks_deleted << " blocks, new genesis is " << genesis_block_id_);
```

Comment:
do you want to print last deleted block also?

Reply:
I think it's redundant as it is obvious that the last deleted is (genesis - 1), if genesis != 1